contracts/src/multisig.cairo

```diff
@@ -0,0 +1,501 @@
+        _set_signers_range(0_usize, signers_len, @signers);
+
+        SignersSet(signers);
+    }
+
+    fn _clean_signers_range(index: usize, len: usize) {
+        if index >= len {
+            return ();
+        }
+
+        let signer = _signers::read(index);
+        _is_signer::write(signer, false);
+        _signers::write(index, Zeroable::zero());
+
+        gas::withdraw_gas_all(get_builtin_costs()).expect('Out of gas');
+        _clean_signers_range(index + 1_usize, len);
+    }
+
+    fn _set_signers_range(index: usize, len: usize, signers: @Array<ContractAddress>) {
+        if index >= len {
+            return ();
+        }
+
+        let signer = *signers.at(index);
+        _signers::write(index, signer);
+        _is_signer::write(signer, true);
+
+        gas::withdraw_gas_all(get_builtin_costs()).expect('Out of gas');
+        _set_signers_range(index + 1_usize, len, signers);
+    }
+
+    fn _get_signers_range(index: usize, len: usize, ref signers: Array<ContractAddress>) {
+        if index >= len {
+            return ();
+        }
+
+        let signer = _signers::read(index);
+        signers.append(signer);
+
+        gas::withdraw_gas_all(get_builtin_costs()).expect('Out of gas');
+        _get_signers_range(index + 1_usize, len, ref signers);
+    }
+
+    fn _set_transaction_calldata_range(
+        nonce: u128, index: usize, len: usize, calldata: @Array<felt252>
+    ) {
+        if index >= len {
+            return ();
+        }
+
+        let calldata_arg = *calldata.at(index);
+        _transaction_calldata::write((nonce, index), calldata_arg);
+
+        gas::withdraw_gas_all(get_builtin_costs()).expect('Out of gas');
+        _set_transaction_calldata_range(nonce, index + 1_usize, len, calldata);
+    }
+
+    fn _get_transaction_calldata_range(
+        nonce: u128, index: usize, len: usize, ref calldata: Array<felt252>
+    ) {
+        if index >= len {
+            return ();
+        }
+
+        let calldata_arg = _transaction_calldata::read((nonce, index));
+        calldata.append(calldata_arg);
+
+        gas::withdraw_gas_all(get_builtin_costs()).expect('Out of gas');
+        _get_transaction_calldata_range(nonce, index + 1_usize, len, ref calldata);
+    }
+
+    fn _set_threshold(threshold: usize) {
+        _threshold::write(threshold);
+        ThresholdSet(threshold);
+    }
+
+    fn _require_signer() {
+        let caller = get_caller_address();
+        let is_signer = _is_signer::read(caller);
+        assert(is_signer, 'invalid signer');
+    }
+
+    fn _require_tx_exists(nonce: u128) {
+        let next_nonce = _next_nonce::read();
+        assert(nonce < next_nonce, 'transaction does not exist');
+    }
+
+    fn _require_not_executed(nonce: u128) {
+        let transaction = _transactions::read(nonce);
+        assert(!transaction.executed, 'transaction already executed');
+    }
+
+    fn _require_not_confirmed(nonce: u128) {
+        let caller = get_caller_address();
```

Comment:
nit: save some code duplication by having one method 

fn _caller_confirmed() -> bool {
      let caller = get_caller_address();
       _is_confirmed::read((nonce, caller));
}

and then no need to define `_require_not_confirmed` and `_require_confirmed` because you can just check the assert value in the other functions